File: Assets/Samples/Unity MCP/1.1.2/Unity MCP Handler Samples/Editor/CodeExecutionCommandHandler.cs
using System;
using System.Collections.Generic;
using Newtonsoft.Json.Linq;
using UnityEditor;
using UnityEngine;
using UnityMCP.Editor.Core;

namespace UnityMCPHandlerSamples.Editor
{
    /// <summary>
    /// Command handler for executing C# code through the MCP interface.
    /// </summary>
    internal sealed class CodeExecutionCommandHandler : IMcpCommandHandler
    {
        /// <summary>
        /// The C# code executer instance.
        /// </summary>
        private readonly CSharpCodeRunner codeCodeRunner;

        /// <summary>
        /// Initializes a new instance of the <see cref="CodeExecutionCommandHandler"/> class.
        /// </summary>
        public CodeExecutionCommandHandler()
        {
            this.codeCodeRunner = new CSharpCodeRunner();
        }

        /// <summary>
        /// Gets the command prefix for this handler.
        /// </summary>
        public string CommandPrefix => "code";

        /// <summary>
        /// Gets the description of this command handler.
        /// </summary>
        public string Description => "Execute C# code in the Unity editor";

        /// <summary>
        /// Executes the command with the given parameters.
        /// </summary>
        /// <param name="action">The action to execute.</param>
        /// <param name="parameters">The parameters for the command.</param>
        /// <returns>A JSON object containing the execution result.</returns>
        public JObject Execute(string action, JObject parameters)
        {
            return action.ToLower() switch
            {
                "execute" => this.ExecuteCode(parameters),
                _ => new JObject
                {
                    ["success"] = false,
                    ["error"] = $"Unknown action: {action}. Supported actions are 'execute' and 'evaluate'."
                }
            };
        }

        /// <summary>
        /// Executes C# code.
        /// </summary>
        /// <param name="parameters">The parameters containing the code to execute.</param>
        /// <returns>A JSON object containing the execution result.</returns>
        private JObject ExecuteCode(JObject parameters)
        {
            var code = parameters["code"]?.ToString();

            if (string.IsNullOrWhiteSpace(code))
            {
                return new JObject
                {
                    ["success"] = false,
                    ["error"] = "The 'code' parameter is required and cannot be empty."
                };
            }

            try
            {
                // Capture logs during execution
                var logMessages = new List<string>();
                var logHandler = new Application.LogCallback((logString, stackTrace, type) =>
                {
                    var prefix = type switch
                    {
                        LogType.Error or LogType.Exception => "[ERROR] ",
                        LogType.Warning => "[WARNING] ",
                        _ => "[INFO] "
                    };
                    logMessages.Add(prefix + logString);
                });

                Application.logMessageReceived += logHandler;

                // Execute on main thread
                this.codeCodeRunner.RunAsync(code);

                // Remove log handler
                Application.logMessageReceived -= logHandler;

                return new JObject
                {
                    ["success"] = true,
                    ["output"] = string.Join("\n", logMessages)
                };
            }
            catch (Exception ex)
            {
                return new JObject
                {
                    ["success"] = false,
                    ["error"] = $"Error executing code: {ex.Message}",
                    ["stackTrace"] = ex.StackTrace
                };
            }
        }
    }
}

File: Packages/jp.shiranui-isuzu.unity-mcp/Samples~/UnityMCPHandlerSamplesJS/CodeExecutionCommandHandler.js
import { z } from "zod";
import { BaseCommandHandler } from "../core/BaseCommandHandler.js";
/**
 * Command handler for executing C# code in the Unity editor.
 */
export class CodeExecutionCommandHandler extends BaseCommandHandler {
    /**
     * Gets the command prefix for this handler.
     */
    get commandPrefix() {
        return "code";
    }
    /**
     * Gets the description of this command handler.
     */
    get description() {
        return "Execute C# code in the Unity editor";
    }
    /**
     * Executes the command with the given parameters.
     * @param action The action to execute.
     * @param parameters The parameters for the command.
     * @returns A Promise that resolves to a JSON object containing the execution result.
     */
    async execute(action, parameters) {
        switch (action.toLowerCase()) {
            case "execute":
                return this.executeCode(parameters);
            default:
                return {
                    success: false,
                    error: `Unknown action: ${action}. Supported actions: execute`
                };
        }
    }
    /**
     * Gets the tool definitions supported by this handler.
     * @returns A map of tool names to their definitions.
     */
    getToolDefinitions() {
        const tools = new Map();
        // Add code_execute tool
        tools.set("code_execute", {
            description: "Execute C# code in the Unity editor",
            parameterSchema: {
                code: z.string().describe("Write direct executable C# code without using statements or class/method wrappers, valid inside a method body, with return statements for value output, and direct access to Unity and .NET APIs.")
            },
            annotations: {
                title: "Execute Code",
                readOnlyHint: false,
                destructiveHint: true,
                idempotentHint: true,
                openWorldHint: false
            }
        });
        return tools;
    }
    /**
     * Executes C# code in the Unity editor.
     * @param parameters Parameters containing the code to execute.
     * @returns A Promise that resolves to a JSON object containing the execution result.
     */
    async executeCode(parameters) {
        try {
            // Validate required parameters
            const code = parameters.code;
            if (!code) {
                return {
                    success: false,
                    error: "The 'code' parameter is required and cannot be empty."
                };
            }
            // First ensure we have a valid connection to Unity
            await this.ensureUnityConnection();
            // Forward the request to Unity
            return await this.sendUnityRequest(`${this.commandPrefix}.execute`, parameters);
        }
        catch (ex) {
            const errorMessage = ex instanceof Error ? ex.message : String(ex);
            console.error(`Error executing code: ${errorMessage}`);
            return {
                success: false,
                error: errorMessage
            };
        }
    }
}
